File: samples/Communication/prize/solution/prize.java
public class prize {
	class Pair{
		int first, second;
		Pair(){
			first = 0;
			second = 0;
		}
		Pair(int x, int y) {
			first = x;
			second = y;
		}
		boolean cmp(Pair p) {
			if(first < p.first) return true;
			if(first > p.first) return false;
			return second < p.second;
		}
	}
	final int MAX = 210000;
	int numb,cnt,ans;
	Pair[] P = new Pair[MAX];
	boolean[] mark = new boolean[MAX];
	public Pair query(int x)
	{
		if(mark[x]) return P[x];
		mark[x]=true;
		int[] result = grader.ask(x);
		if(result[0]+result[1]==0) ans=x;
		return P[x]=new Pair(result[0],result[1]);
	}

	public void bs(int l,int r,int nl,int nr)
	{
		if((cnt--)<=0) return;
		if(l>r) return;
		for(int i=0;i<=r-l;i++)
		{
			int mid,midl=(l+r)/2-i/2,midr=(l+r)/2+(i+1)/2;
			if(i%2==0) mid=midl;
			else mid=midr;
			Pair tmp=query(mid);
			if(tmp.first+tmp.second>numb) {cnt=0;return;}
			if(tmp.first+tmp.second==numb)
			{
				int tmpl=(i%2==0?0:midr-midl);
				int tmpr=(i%2==1?0:midr-midl);
				if(tmp.first-tmpl>nl) bs(l,midl-1,nl,tmp.second+tmpl);
				if(tmp.second-tmpr>nr) bs(midr+1,r,tmp.first+tmpr,nr);
				break;
			}
		}
	}

	public int find_best(int n)
	{
		if(n==1) return 0;
		numb=1;
		cnt=20;
		bs(0,n-1,0,0);
		int p=0;
		for(int i=0;i<Math.sqrt(n)+30 && i<n && numb<=26;i++)
		{
			Pair tmp=query(i);
			if(tmp.first+tmp.second>numb) p=i;
			numb=Math.max(numb,tmp.first+tmp.second);
		}
		cnt=1000000;
		bs(p,n-1,p,0);
		return ans;
	}
}

File: samples/Communication/prize/solution/prize.pas
{
	Author: Saeed
	Verdict: Correct solution
}

unit prize;

interface

type
	TIntArray = array of longint;

function find_best(n : longint) : longint;


implementation

uses graderlib;

var
    cnt : longint;

function bs(l, r, nl, nr, numb : longint) : longint;
var
	tmp : TIntArray;
	i, tmpl, tmpr, midl, midr, mid : longint;
begin
	if cnt <= 0 then
	begin
		bs := -1;
		exit;
	end;
        dec(cnt);
	if l > r then
	begin
		bs := -1;
		exit;
	end;
	for i := 0 to r-l do
	begin
		midl := (l + r) div 2 - i div 2;
		midr := (l + r) div 2 + (i+1) div 2;
		if i mod 2 = 0 then
			mid := midl
		else
			mid := midr;
		tmp := ask(mid);
		if tmp[0] + tmp[1] = 0 then
		begin
			bs := mid;
			exit;
		end;
                if tmp[0] + tmp[1] > numb then
                begin
                    cnt := 0;
                    bs := -1;
                    exit;
                end;
		if tmp[0] + tmp[1] = numb then
		begin
			if i mod 2 = 0 then
                        begin
				tmpl := 0;
				tmpr := midr - midl;
                        end
			else
                        begin
				tmpl := midr - midl;
				tmpr := 0;
			end;
			if tmp[0] - tmpl > nl then
			begin
				bs := bs(l, midl-1, nl, tmp[1] + tmpl, numb);
				if bs <> -1 then
					exit;
			end;
			if tmp[1] - tmpr > nr then
                        begin
				bs := bs(midr+1, r, tmp[0] + tmpr, nr, numb);
				if bs <> -1 then
					exit;
                        end;
                        break;
		end;
	end;
	bs := -1;
end;

function find_best(n : longint) : longint;
var
	tmp : TIntArray;
	p, i, numb : longint;
begin
	if n = 1 then
	begin
		find_best := 0;
		exit;
	end;
	numb := 1;
        cnt := 20;
	find_best := bs(0, n-1, 0, 0, numb);
	if find_best <> -1 then
		exit;
	p := 0;
	for i := 0 to round(Sqrt(n)) + 30 do
	begin
		if numb > 26 then
			break;
                tmp := ask(i);
		if tmp[0] + tmp[1] = 0 then
		begin
			find_best := i;
			exit;
		end;
		if numb < tmp[0] + tmp[1] then
                begin
                        writeln(p);
			p := i;
			numb := tmp[0] + tmp[1];
                end;
	end;
        cnt := 1000000;
	find_best := bs(p, n-1, p, 0, numb);
end;

end.
